### ISSUE TITLE
unable to install flash_attn

### ISSUE THREAD

**Issue by hulululu0725** [jzhang38/TinyLlama#201]
![Image](https://github.com/user-attachments/assets/90c7b287-b82e-47ff-80d3-b60f939469ae)
output of pip list:
Package               Version
--------------------- ------------------
accelerate            0.21.0
adjustText            0.8
aiohappyeyeballs      2.4.4
aiohttp               3.11.11
aiosignal             1.3.2
annotated-types       0.7.0
ansicon               1.89.0
anyio                 4.6.2.post1
appdirs               1.4.4
arrow                 1.3.0
async-timeout         5.0.1
attrs                 24.3.0
backoff               2.2.1
beautifulsoup4        4.12.3
blessed               1.19.1
boto3                 1.36.1
botocore              1.36.1
Brotli                1.1.0
calmap                0.0.11
certifi               2023.11.17
cffi                  1.17.1
charset-normalizer    3.3.2
click                 8.1.8
colorama              0.4.6
colourmap             1.1.16
contourpy             1.3.1
croniter              1.4.1
cycler                0.12.1
dataclasses           0.8
datasets              2.14.4
datazets              0.1.9
dateutils             0.6.12
deepdiff              7.0.1
dill                  0.3.7
distro                1.9.0
dnspython             2.7.0
docker-pycreds        0.4.0
editor                1.6.6
einops                0.8.0
email_validator       2.2.0
et-xmlfile            1.1.0
eval_type_backport    0.2.2
exceptiongroup        1.2.2
fastapi               0.115.6
fastapi-cli           0.0.7
filelock              3.13.3
flash-attn            2.3.2
fonttools             4.55.3
frozenlist            1.5.0
fsspec                2024.10.0
gitdb                 4.0.12
GitPython             3.1.44
h11                   0.14.0
h2                    4.1.0
hpack                 4.0.0
httpcore              1.0.6
httptools             0.6.4
httpx                 0.27.2
huggingface_hub       0.27.1
hyperframe            6.0.1
idna                  3.4
importlib_metadata    8.5.0
inquirer              3.4.0
itsdangerous          2.2.0
Jinja2                3.1.4
jinxed                1.2.1
jiter                 0.6.1
jmespath              1.0.1
joblib                1.3.2
jsonargparse          4.7.3
kiwisolver            1.4.8
lightning             2.5.0.post0
lightning_cloud       0.5.70
lightning-utilities   0.11.9
markdown-it-py        3.0.0
MarkupSafe            2.1.5
matplotlib            3.10.0
mdurl                 0.1.2
mizani                0.10.0
mpmath                1.3.0
multidict             6.1.0
multiprocess          0.70.15
networkx              3.3
ninja                 1.11.1.3
numpy                 1.26.4
openai                1.52.0
openpyxl              3.1.2
ordered-set           4.1.0
orjson                3.10.14
packaging             24.2
pandas                2.2.2
patsy                 0.5.3
pca                   2.0.5
pillow                11.1.0
pip                   24.1.2
platformdirs          4.3.6
plotnine              0.12.0
propcache             0.2.1
protobuf              5.28.3
psutil                6.1.1
pyarrow               19.0.0
pycparser             2.22
pydantic              2.9.2
pydantic_core         2.23.4
Pygments              2.19.1
PyJWT                 2.10.1
pyparsing             3.2.1
PySocks               1.7.1
python-dateutil       2.9.0.post0
python-dotenv         1.0.1
python-multipart      0.0.20
pytorch-lightning     2.5.0.post0
pytz                  2024.2
PyYAML                6.0.2
readchar              4.2.0
regex                 2024.11.6
requests              2.31.0
rich                  13.9.4
rich-toolkit          0.11.3
runs                  1.2.2
s3transfer            0.11.1
sacremoses            0.0.53
safetensors           0.5.2
scatterd              1.3.7
scikit-learn          1.3.2
scipy                 1.11.4
seaborn               0.13.0
sentencepiece         0.2.0
sentry-sdk            2.20.0
setproctitle          1.3.4
setuptools            75.8.0
shellingham           1.5.4
six                   1.17.0
smmap                 5.0.0
sniffio               1.3.1
soupsieve             2.5
starlette             0.41.3
starsessions          1.3.0
statsmodels           0.14.0
sympy                 1.13.1
threadpoolctl         3.2.0
tokenizers            0.13.3
torch                 2.1.0+cu118
torchaudio            2.1.0+cu118
torchmetrics          1.6.1
torchvision           0.16.0+cu118
tqdm                  4.66.5
traitlets             5.14.3
transformers          4.31.0
typer                 0.15.1
typer-slim            0.15.1
types-python-dateutil 2.9.0.20241206
typing_extensions     4.11.0
tzdata                2024.2
urllib3               2.1.0
uvicorn               0.34.0
wandb                 0.19.4
watchfiles            1.0.4
wcwidth               0.2.13
websocket-client      1.8.0
websockets            12.0
wheel                 0.41.3
win_inet_pton         1.1.0
xformers              0.0.22.post4+cu118
xmod                  1.8.1
xxhash                3.5.0
yarl                  1.18.3
zipp                  3.21.0

1. zstandard             0.23.0

**Comment by shrey-mishra** [comment]
![Image](https://github.com/user-attachments/assets/2d8324f0-777e-403e-a9e9-07fd4d675029)

I encountered a similar issue where I was unable to proceed with tokenizing my dataset. I found that this can be resolved using the `AutoTokenizer` from the `transformers` library. To streamline the process, I created a script called `prepare_csv_data.py`, which extracts key features relevant to my use case.

Code which I used in `prepare_csv_data.py`.


```import pandas as pd
import argparse
import os
import torch
from transformers import AutoTokenizer
from sklearn.preprocessing import LabelEncoder

def tokenize_and_save(source_path, tokenizer_path, destination_path, text_column, split):
    # Load CSV file
    df = pd.read_csv("/content/traffic_sample_1year_with_packet_length.csv")

    # Convert categorical columns to tokens
    ip_encoder = LabelEncoder()
    df["Source IP"] = ip_encoder.fit_transform(df["Source IP"])
    df["Destination IP"] = ip_encoder.fit_transform(df["Destination IP"])

    # Convert Protocol and Direction to numerical tokens
    df["Protocol"] = df["Protocol"].astype("category").cat.codes
    df["Direction"] = df["Direction"].astype("category").cat.codes

    # Ensure the specified text column exists
    if text_column not in df.columns:
        raise ValueError(f"Column '{text_column}' not found in CSV. Available columns: {df.columns}")

    # Keep numerical features (Timestamp, Packet Length, and other numerical fields)
    numerical_features = df[["Timestamp", "Packet Length"]].values

    # Combine features into a text format for tokenization
    df["text"] = df.apply(lambda row: f"{row['Source IP']} {row['Destination IP']} {row['Protocol']} {row['Direction']} {row[text_column]}", axis=1)

    # Load tokenizer
    tokenizer = AutoTokenizer.from_pretrained(tokenizer_path)

    # Tokenize text-based features
    tokenized_texts = tokenizer(df["text"].tolist(), padding=True, truncation=True, max_length=512, return_tensors="pt")

    # Save tokenized data
    os.makedirs(destination_path, exist_ok=True)
    tokenized_output_path = os.path.join(destination_path, f"{split}_tokenized.pt")

    # Save both tokenized categorical and numerical features
    torch.save({"tokens": tokenized_texts, "numerical": numerical_features}, tokenized_output_path)

    print(f"Tokenized data saved to {tokenized_output_path}")

if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument("--source_path", type=str, required=True, help="Path to CSV file")
    parser.add_argument("--tokenizer_path", type=str, required=True, help="Path to tokenizer")
    parser.add_argument("--destination_path", type=str, required=True, help="Where to save tokenized data")
    parser.add_argument("--text_column", type=str, required=True, help="Text column to tokenize")  # ✅ Added this line
    parser.add_argument("--split", type=str, required=True, help="Dataset split (train/validation/test)")

    args = parser.parse_args()

    tokenize_and_save(args.source_path, args.tokenizer_path, args.destination_path, args.text_column, args.split)```
